File: 3217/Solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func modifiedList(nums []int, head *ListNode) *ListNode {
    max := 0
    for _, num := range nums {
        if max < num {
            max = num
        }
    }
    seen := make([]int, max + 1)

    for _, num := range nums {
        seen[num]++
    }

    temp := &ListNode{Val: 0, Next: head}
    cur := temp
    for cur.Next != nil {
        if cur.Next.Val <= max && seen[cur.Next.Val] > 0{
            cur.Next = cur.Next.Next
        }else{
            cur = cur.Next
        }
    }
    return temp.Next
}
// func modifiedList(nums []int, head *ListNode) *ListNode {
//     seen := make(map[int]bool)
//     for _, num := range nums {
//         seen[num] = true
//     }
//     temp := &ListNode{Val: 0, Next: head}
//     cur := temp
//     for cur.Next != nil {
//         if seen[cur.Next.Val] {
//             cur.Next = cur.Next.Next
//         }else{
//             cur = cur.Next
//         }
//     }
//     return temp.Next
// }

File: 3217/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        int max = 0;
        for(int i : nums) if(max < i) max = i;

        boolean[] seen = new boolean[max + 1];
        
        for(int i : nums) seen[i] = true;

        ListNode dummy = new ListNode(0, head);
        ListNode cur = dummy;
        while(cur.next != null){
            if( cur.next.val <= max && seen[cur.next.val]){
                cur.next = cur.next.next;
            }else{
                cur = cur.next;
            }
        }
        return dummy.next;
    }
}

// time O(n)
// space O(max)
